modules/gateway/nodes.go

```diff
@@ -43,17 +39,28 @@ func (g *Gateway) removeNode(addr modules.NetAddress) error {
 	return nil
 }
 
+// randomNode returns a random node from the gateway. An error can be returned
+// if there are no nodes in the node list.
 func (g *Gateway) randomNode() (modules.NetAddress, error) {
-	if len(g.nodes) > 0 {
-		r, _ := crypto.RandIntn(len(g.nodes))
-		for node := range g.nodes {
-			if r <= 0 {
-				return node, nil
-			}
-			r--
-		}
+	if len(g.nodes) == 0 {
+		return "", errNoPeers
 	}
 
+	// Select a random peer. Note that the algorithm below is roughly linear in
+	// the number of nodes known by the gateway, and this number can approach
+	// every node on the network. If the network gets large, this algorithm
+	// will either need to be refactored, or more likely a cap on the size of
+	// g.nodes will need to be added.
```

Comment:
I bet there's some horrible (i.e. reflection-based) hack that could return a cryptographically random map value in constant time.
At first I thought you could use `reflect.ValueOf(myMap).MapKeys()`, since it returns a slice (which you can randomly index). But in fact this is even slower than what we're doing now, because the [implementation](https://golang.org/src/reflect/value.go?s=32392:32424#L1052) iterates through all the map keys _and_ allocates a new slice for them.
So you'd probably have to do something **really** gross involving unsafe access to the map's buckets :P